File: day-4/four.js
const min = 128392;
const max = 643281;
let current = min;

function hasNoDecreasingDigits(values) {
  let valid = true;

  values.forEach((value, i) => {
    if (i !== 0 && value < values[i - 1]) {
      valid = false;
    }
  });

  return valid;
}

function hasAdjacentEqualDigits(values) {
  let valid = false;

  values.forEach(password => {
    password.forEach((digit, i) => {
      if (
        i !== 0 &&
        digit === password[i - 1] &&
        digit !== password[i - 2] &&
        digit !== password[i + 1]
      ) {
        valid = true;
      }
    });
  });

  return valid;
}

function checkForPasswords() {
  const noDecendingPasswords = [];

  while (current <= max) {
    const values = String(current).split('');
    if (hasNoDecreasingDigits(values)) {
      noDecendingPasswords.push([values]);
    }

    current++;
  }

  const validPasswords = noDecendingPasswords.filter(hasAdjacentEqualDigits);

  console.log(validPasswords.length);
}

checkForPasswords();

File: day-4/four.rb
def has_no_decreasing(password) 
  valid = true

  password.each_with_index do |digit, i|
    if i != 0 && digit < password[i -1]
      valid = false
    end
  end

  valid 
end

def has_adjacent_equal_digits(password) 
  valid = false

  password.each_with_index do |digit, i|
    if 
      i != 0 &&
      digit == password[i - 1] &&
      digit != password[i - 2] &&
      digit != password[i + 1] 
      valid = true
    end
  end

  valid 
end

valid_passwords = []

(128392..643281).each do |password|
  password_array = password.to_s.split('') 
  if has_no_decreasing(password_array) && has_adjacent_equal_digits(password_array)
    valid_passwords.push(password)
  end
end

puts valid_passwords.length